src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -5943,14 +5911,33 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
     }
 
     def typed(tree: Tree, mode: Mode, pt: Type): Tree = {
+      // Extract and insert stabilizing ValDefs (if any) which might have been
+      // introduced during the typing of the original expression.
+      @inline def insertStabilizer(doTyped: Mode => Tree): Tree = {
+        if (phase.erasedTypes || mode.in(APPSELmode) || isMacroImplRef(tree)) doTyped(mode)
+        else tree match {
+          case _: Select | _: Apply | _: TypeApply =>
+            val saved = context.pendingStabilizers
+            context.pendingStabilizers = Nil
+            val res = doTyped(mode | APPSELmode)
+            val vdefs = context.pendingStabilizers
+            context.pendingStabilizers = saved
+            if (vdefs.isEmpty)
+              res
+            else
+              typed1(Block(vdefs.reverse, res) setPos res.pos, mode, pt)
```

Comment:
`assert(vdefs.forall(_.owner == context.owner), (vdefs.map(sym => (sym, sym.owner)), context.owner)`

(Yeah, I've been chasing owner corruption bugs today...)

Reply:
Ran the test suite with this assertion, but for the release I guess it's better to do without it.

Reply:
`devWarningIf(...)` instead?